protocol/communities_messenger_test.go

```diff
@@ -3684,3 +3684,170 @@ func (s *MessengerCommunitiesSuite) TestRetrieveBigCommunity() {
+
+	// Make sure clock is not empty
+	s.Require().NotEmpty(requestToJoin1.Clock)
+
+	s.Require().Len(response.Communities(), 1)
+	s.Require().Equal(response.Communities()[0].RequestedToJoinAt(), requestToJoin1.Clock)
+
+	// pull all communities to make sure we set RequestedToJoinAt
+
+	allCommunities, err := s.alice.Communities()
+	s.Require().NoError(err)
+	s.Require().Len(allCommunities, 2)
+
+	if bytes.Equal(allCommunities[0].ID(), community.ID()) {
+		s.Require().Equal(allCommunities[0].RequestedToJoinAt(), requestToJoin1.Clock)
+	} else {
+		s.Require().Equal(allCommunities[1].RequestedToJoinAt(), requestToJoin1.Clock)
+	}
+
+	// pull to make sure it has been saved
+	requestsToJoin, err := s.alice.MyPendingRequestsToJoin()
+	s.Require().NoError(err)
+	s.Require().Len(requestsToJoin, 1)
+
+	// Make sure the requests are fetched also by community
+	requestsToJoin, err = s.alice.PendingRequestsToJoinForCommunity(community.ID())
+	s.Require().NoError(err)
+	s.Require().Len(requestsToJoin, 1)
+
+	requestToJoin2 := response.RequestsToJoinCommunity[0]
+
+	s.Require().NotNil(requestToJoin2)
+	s.Require().Equal(community.ID(), requestToJoin2.CommunityID)
+	s.Require().NotEmpty(requestToJoin2.ID)
+	s.Require().NotEmpty(requestToJoin2.Clock)
+	s.Require().Equal(requestToJoin2.PublicKey, common.PubkeyToHex(&s.alice.identity.PublicKey))
+	s.Require().Equal(communities.RequestToJoinStatePending, requestToJoin2.State)
+
+	s.Require().Equal(requestToJoin1.ID, requestToJoin2.ID)
+
+	// Cancel request to join community
+	requestsToJoin, err = s.alice.MyPendingRequestsToJoin()
+	s.Require().NoError(err)
+	s.Require().Len(requestsToJoin, 1)
+
+	requestToJoin := requestsToJoin[0]
+
+	requestToCancel := &requests.CancelRequestToJoinCommunity{ID: requestToJoin.ID}
+	response, err = s.alice.CancelRequestToJoinCommunity(ctx, requestToCancel)
+	s.Require().NoError(err)
+	s.Require().NotNil(response)
+	s.Require().Len(response.RequestsToJoinCommunity, 1)
+	s.Require().Equal(communities.RequestToJoinStateCanceled, response.RequestsToJoinCommunity[0].State)
+
+	// pull to make sure it has been saved
+	cancelRequestsToJoin, err := s.alice.MyCanceledRequestsToJoin()
+	s.Require().NoError(err)
+	s.Require().Len(cancelRequestsToJoin, 1)
+	s.Require().Equal(cancelRequestsToJoin[0].State, communities.RequestToJoinStateCanceled)
+
+	// Retrieve cancel request to join to owner
+	err = tt.RetryWithBackOff(func() error {
+		response, err = s.owner.RetrieveAll()
+		if err != nil {
+			return err
+		}
+		if len(response.RequestsToJoinCommunity) == 0 {
+			return errors.New("request to join community not received")
+		}
+		return nil
+	})
+
+	s.Require().NoError(err)
+	var response2 *MessengerResponse
+	// Fetch new messages for Alice
+	err = tt.RetryWithBackOff(func() error {
+		response2, err = s.alice.RetrieveAll()
+		if err != nil {
+			return err
+		}
+		if len(response2.Communities()) == 0 {
+			return errors.New("communities not received")
+		}
```

Comment:
There are no guarantees in which order community updates will be retrieved by alice. This may be the first update where alice is still a member of a community.